server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java

```diff
@@ -0,0 +1,165 @@
+      var argTimeout = ArgumentUtils.toDouble(arguments.get(lastKeyIdx));
+      if (argTimeout < 0) {
+         RespErrorUtil.mustBePositive(handler.allocator());
+         return handler.myStage();
+      }
+      long timeout = (long) (argTimeout * Duration.ofSeconds(1).toMillis());
+      // If all the keys are empty or null, create a listener for each key
+      // otherwise return the left value of the first non empty list
+      var pollStage = pollAllKeys(listMultimap, arguments);
+      // If no value returned, we need subscribers
+      return pollStage.thenCompose(v -> handler.stageToReturn(
+            (v != null && !v.isEmpty())
+                  ? CompletableFuture.completedFuture(v)
+                  : addSubscriber(listMultimap, filterKeys, timeout, handler, ctx),
+            ctx, Consumers.COLLECTION_BULK_BICONSUMER));
+   }
+
+   CompletionStage<Collection<byte[]>> pollKeyValue(EmbeddedMultimapListCache<byte[], byte[]> mmList, byte[] key) {
+      return mmList.pollFirst(key, 1)
+            .thenApply((v) -> (v == null || v.isEmpty())
+                  ? null
+                  : Arrays.asList(key, v.iterator().next()));
+   }
+
+   CompletionStage<Collection<byte[]>> addSubscriber(EmbeddedMultimapListCache<byte[], byte[]> listMultimap,
+         List<byte[]> filterKeys, long timeout, Resp3Handler handler,
+         ChannelHandlerContext ctx) {
+      if (log.isTraceEnabled()) {
+         log.tracef("Subscriber for keys: " +
+               filterKeys.toString());
+      }
+      AdvancedCache<Object, Object> cache = handler.cache().withMediaType(MediaType.APPLICATION_OCTET_STREAM, null);
+      DataConversion vc = cache.getValueDataConversion();
+      PubSubListener pubSubListener = new PubSubListener(ctx.channel());
+      EventListenerKeysFilter filter = new EventListenerKeysFilter(filterKeys.toArray(byte[][]::new),
+            handler.cache().getKeyDataConversion());
+      CacheEventConverter<Object, Object, Object> converter = (key, oldValue, oldMetadata, newValue, newMetadata,
+            eventType) -> vc.fromStorage(newValue);
+      var listnerStage = cache.addListenerAsync(pubSubListener, filter, converter);
+
+      // start a timer if needed
+      final ScheduledFuture<?> scheduledTimer = (timeout > 0) ?
+         ctx.channel().eventLoop().schedule(() -> {
+            pubSubListener.getFuture().complete(null);
+         }, timeout, TimeUnit.MILLISECONDS) : null;
+      // Listener can lose events during its install, so we need to poll again
+      // and if we get values complete the listener future.
+      listnerStage.thenCompose(v -> pollAllKeys(listMultimap,filterKeys)).thenAccept(v -> {
+         if (v != null) {
+            pubSubListener.getFuture().complete(v);
+            if (scheduledTimer != null) {
+               scheduledTimer.cancel(true);
+            }
+         }
+      });
+      // Uninstall the listener when got values
+      return pubSubListener.getFuture().thenApply(v -> {
```

Comment:
Sorry, I didn't notice at first. This would never finish if the `listnerStage` fails to install the listener. I think if the listener fails to install, we can simply return an error to the user. WDYT?

Reply:
> Sorry, I didn't notice at first. This would never finish if the `listnerStage` fails to install the listener. I think if the listener fails to install, we can simply return an error to the user. WDYT?

sonds good, I added some lines for that

Reply:
I see the try-catch above, only that wouldn't be enough. I think we can use the `CompletionStages.handleAndCompose` here to handle the failures of the listener completable.

Reply:
Yeah, the problem is caused because we don't return the `listnerStage,` only the `pubSubListener.getFuture().`. To test that case we need to use mocks, got something that simulates the problem:

```java
   public void testBlpopFailsInstallingListener() throws Exception {
      RedisCommands<String, String> redis = redisConnection.sync();
      var client = createClient(30000, server.getPort());
      RedisAsyncCommands<String, String> redisAsync = client.connect().async();

      CacheNotifierImpl<?, ?> cni = (CacheNotifierImpl<?, ?>) TestingUtil.extractComponent(cache, CacheNotifier.class);
      CacheNotifierImpl<?, ?> spyCni = spy(cni);
      CountDownLatch latch = new CountDownLatch(1);

      Answer<CompletionStage<Void>> listenerAnswer = ignore -> {
         latch.countDown();
         return CompletableFuture.failedFuture(new RuntimeException("Injected failure"));
      };

      if (simpleCache) doAnswer(listenerAnswer).when(spyCni).addListenerAsync(any(), any(), any());
      else doAnswer(listenerAnswer).when(spyCni).addListenerAsync(any(), any(), any(), any());

      try {
         TestingUtil.replaceComponent(cache, CacheNotifier.class, spyCni, true);

         CompletableFuture<KeyValue<String, String>> blpop = redisAsync.blpop(0, "my-nice-key")
               .toCompletableFuture();
         assertThat(latch.await(10, TimeUnit.SECONDS)).isTrue();
         redis.rpush("my-nice-key", "v1", "v2", "v3");
         KeyValue<String, String> res = blpop.get(10, TimeUnit.SECONDS);

         assertThat(res.getKey()).isEqualTo("my-nice-key");
         assertThat(res.getValue()).isEqualTo("v1");
      } finally {
         RespTestingUtil.killClient(client);
         TestingUtil.replaceComponent(cache, CacheNotifier.class, cni, true);
      }
   }
```

This test will fail with a timeout exception because the BLPOP operation never completes. We can catch the exception with the `CompletionStages.handleAndCompose` and complete the `pubSubListener.getFuture().completeExceptionally`. So that stageToReturn() automatically handles it.

Reply:
I'll also check later if we are handling this kind of exception from installing listeners in other commands.

Reply:
@jabolina thanks for the test case!
I partially rewrote the implementation changing to `addListener` from `addListenerAsync` because the async version hangs at the end of `perform()` and I was unable to find the cause. `addListener` is executed in a stage so should be non-blocking. I kept the commit history if we want to rollback to async.
pls have a look